spark/src/main/java/org/apache/iceberg/spark/source/Writer.java

```diff
@@ -303,96 +305,179 @@ public String toString() {
+      // That's useful, for example, if a Spark job dies and leaves files in the file system, you can identify them all
+      // with a recursive listing and grep.
+      private final String uuid = UUID.randomUUID().toString();
+      private int fileCount;
+
+      EncryptedOutputFileFactory(int partitionId, long taskId, long epochId) {
+        this.partitionId = partitionId;
+        this.taskId = taskId;
+        this.epochId = epochId;
+        this.fileCount = 0;
+      }
+
+      private synchronized String generateFilename() {
+        return format.addExtension(String.format("%05d-%d-%s-%05d", partitionId, taskId, uuid, fileCount++));
+      }
+
+      /**
+       * Generates EncryptedOutputFile for UnpartitionedWriter.
+       */
+      public EncryptedOutputFile newOutputFile() {
+        OutputFile file = fileIo.newOutputFile(locations.newDataLocation(generateFilename()));
+        return encryptionManager.encrypt(file);
+      }
+
+      /**
+       * Generates EncryptedOutputFile for PartitionedWriter.
+       */
+      public EncryptedOutputFile newOutputFile(PartitionKey key) {
+        OutputFile rawOutputFile = fileIo.newOutputFile(locations.newDataLocation(spec, key, generateFilename()));
+        return encryptionManager.encrypt(rawOutputFile);
+      }
+    }
   }
 
   private interface AppenderFactory<T> {
     FileAppender<T> newAppender(OutputFile file, FileFormat format);
   }
 
-  private static class UnpartitionedWriter implements DataWriter<InternalRow>, Closeable {
-    private final FileIO fileIo;
-    private FileAppender<InternalRow> appender = null;
-    private Metrics metrics = null;
-    private List<Long> offsetRanges = null;
-    private final EncryptedOutputFile file;
+  private interface OutputFileFactory<T> {
+    T newOutputFile();
+    T newOutputFile(PartitionKey key);
+  }
 
-    UnpartitionedWriter(
-        EncryptedOutputFile outputFile,
-        FileFormat format,
-        AppenderFactory<InternalRow> factory,
-        FileIO fileIo) {
+  @SuppressWarnings("checkstyle:VisibilityModifier") // direct access desired from sub-classes for performance.
+  private abstract static class BaseWriter implements DataWriter<InternalRow> {
+    protected static final int ROWS_DIVISOR = 1000;
+
+    protected final Set<PartitionKey> completedPartitions = Sets.newHashSet();
+    protected final List<DataFile> completedFiles = Lists.newArrayList();
+    protected final PartitionSpec spec;
+    protected final FileFormat format;
+    protected final AppenderFactory<InternalRow> appenderFactory;
+    protected final OutputFileFactory<EncryptedOutputFile> fileFactory;
+    protected final PartitionKey key;
+    protected final FileIO fileIo;
+    protected final long targetFileSize;
+    protected PartitionKey currentKey = null;
+    protected FileAppender<InternalRow> currentAppender = null;
+    protected EncryptedOutputFile currentFile = null;
+    protected long currentRows;
```

Comment:
Nit: all fields should be initialized.

Reply:
It is also strange that this is only used in child classes.

Can this class provide a `writeInternal` method that updates `currentRows`, writes to the appender, and checks when to open and close the current appender? That would be cleaner and would no longer require all the protected fields. Does performance really degrade when the subclasses are a bit more separated from this base class?

Reply:
I'd still need field `currentKey` to be accessible, since it is used in `openCurrent()` and `closeCurrent()` in base class, but also on `PartitionedWriter`'s `write()`.

So I can:
1) Only make `currentKey` protected.
or
2) Add getter/setter for this particular field.

WDYT?

> Does performance really degrade when the subclasses are a bit more separated from this base class?

A method call is more expensive than a field access, although I admit that the JIT compiler should pick this up right away. So will fix.

Reply:
I'd say to use the getter/setter.

I think this commit is almost ready, except for the encapsulation in this class. If we think the JIT compiler will handle this, then let's go ahead and make this base class handle the important parts.